File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def bfs(start, computers, visited):
    q = deque([start])
    visited[start] = 1
    
    while q:
        cur = q.popleft()
        
        for idx in range(len(computers[cur])):
            # next = computers[cur][idx]
            # print(cur, next)
            if visited[idx] == 1:
                continue
            if computers[cur][idx] == 1:
                q.append(idx)
                visited[idx] = 1
    
def solution(n, computers):
    answer = 0
    visited = [0 for _ in range(n)]
    
    for i in range(n):
        if visited[i] == 0:
            bfs(i, computers, visited)
            print(visited)
            answer += 1
            
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

void bfs(int start, const vector<vector<int>>& computers, vector<bool>& visited) {
    queue<int> q;
    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int cur = q.front();
        q.pop();

        for (int i = 0; i < computers[cur].size(); ++i) {
            if (!visited[i] && computers[cur][i] == 1) {
                visited[i] = true;
                q.push(i);
            }
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    vector<bool> visited(n, false);

    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            bfs(i, computers, visited);
            ++answer;
        }
    }

    return answer;
}